File: Hash/149.Max-Points-on-a-Line/149.Max-Points-on-a-Line.cpp
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    int maxPoints(vector<Point>& points) 
    {
        int result=0;
        
        for (int i=0; i<points.size(); i++)
        {
            unordered_map <int,int> Map;
            int countSame=0;
            
            for (int j=0; j<points.size(); j++)
            {
                if (j==i) continue;
                
                if (points[i].x==points[j].x && points[i].y==points[j].y )
                {
                    countSame++;
                    continue;
                }
                
                int flag=0;
                for (auto a:Map)
                {
                    Point p=points[a.first];
                    if ( (long long)(p.y-points[i].y)*(points[j].x-points[i].x) == (long long)(points[j].y-points[i].y)*(p.x-points[i].x) )
                    {
                        flag=1;
                        Map[a.first]++;
                        break;
                    }
                }
                
                if (flag==0)                
                    Map[j]=1;                               
            }
            
            int temp=0;
            for (auto a:Map)
                temp = max(temp,a.second);
            temp+=(countSame+1);
            result=max(result,temp);
        }
        
        return result;
    }
};

File: Hash/149.Max-Points-on-a-Line/149.Max-Points-on-a-Line.py
# Definition for a point.
# class Point:
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution:
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """ 
        
        def inline(a,b,c):
            return (b.y-a.y)*(c.x-a.x)==(c.y-a.y)*(b.x-a.x)
                        
        if (len(points)<=2):
            return len(points)
        
        result = 0;
        for i in range(len(points)):                        
            D = collections.defaultdict(int);
            countSame = 0;            
            for j in range(len(points)):                
                if (j==i): 
                    continue;                         
                if points[j].x==points[i].x and points[j].y==points[i].y: 
                    countSame+=1;
                    continue;
                flag = 0;
                for key in D:                   
                    if inline(points[i],points[key],points[j])==1:
                        flag = 1;
                        D[key]+=1
                        break;
                if (flag==0):
                    D[j]=1
            result = max(result, max(list(D.values())+[0])+countSame+1)
            
        return result
